pkg/localclusterimport/import_local_cluster.go

```diff
@@ -0,0 +1,240 @@
+package localclusterimport
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	configv1 "github.com/openshift/api/config/v1"
+	hiveext "github.com/openshift/assisted-service/api/hiveextension/v1beta1"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/apis/hive/v1/agent"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const LocalClusterNamespace = "local-cluster"
+
+type LocalClusterImport struct {
+	clusterImportOperations ClusterImportOperations
+	log                     *logrus.Logger
+}
+
+func NewLocalClusterImport(localClusterImportOperations ClusterImportOperations, log *logrus.Logger) LocalClusterImport {
+	return LocalClusterImport{clusterImportOperations: localClusterImportOperations, log: log}
+}
+
+func (i *LocalClusterImport) shouldImportLocalCluster() (bool, error) {
+	// The presence of agentClusterInstall for the local cluster will determine whether or not we need to perform an import
+	agentClusterInstall, err := i.clusterImportOperations.GetAgentClusterInstall(LocalClusterNamespace, LocalClusterNamespace+"-cluster-install")
+	if err != nil && !k8serrors.IsNotFound(err) {
+		i.log.Errorf("hubClusterDay2Importer: There was an error fetching the AgentClusterInstall while checking for need to register local cluster.")
+		return false, err
+	}
+	if agentClusterInstall != nil {
+		i.log.Infof("hubClusterDay2Importer: Found AgentClusterInstall for hub cluster, assuming that hub has been correctly registered for ZTP day 2 operations: %s", agentClusterInstall.Name)
+		return false, nil
+	}
+	return true, nil
+}
+
+func (i *LocalClusterImport) ImportLocalCluster() error {
+	clusterVersion, err := i.clusterImportOperations.GetClusterVersion("version")
+	if err != nil {
+		i.log.Errorf("hubClusterDay2Importer: Unable to find cluster version info due to error: %s", err.Error())
+		return err
+	}
+
+	shouldImportLocalCluster, err := i.shouldImportLocalCluster()
+	if err != nil {
+		return err
+	}
+	if !shouldImportLocalCluster {
+		message := "hubClusterDay2Importer: no need to import local cluster as registration already detected"
+		i.log.Info(message)
+		return errors.New(message)
+	}
+	openshift_version := ""
+	release_image := ""
+	if clusterVersion != nil && clusterVersion.Status.History[0].State == configv1.CompletedUpdate {
```

Comment:
is it always the first?

Reply:
Seems that way, yes, the State is the crucial thing, it will be `configv1.PartialUpdate` until the installation has completed.
But yes, I tested this assumption using `oc adm upgrade` -- the first history item is always the latest

Reply:
can you install older ocp and upgrade to make sure it will remain the first one?

Reply:
I will try this now to verify

Reply:
I can confirm that the latest version is always the [0]th entry

`        "history": [
            {
                "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.13.11-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nPrecondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=True (), so the update from 4.13.9 to 4.13.11 is probably neither recommended nor supported.",
                "completionTime": null,
                "image": "quay.io/openshift-release-dev/ocp-release:4.13.11-x86_64",
                "startedTime": "2023-09-06T08:52:43Z",
                "state": "Partial",
                "verified": false,
                "version": "4.13.11"
            },
            {
                "completionTime": "2023-09-06T08:52:09Z",
                "image": "quay.io/openshift-release-dev/ocp-release@sha256:a266d3d65c433b460cdef7ab5d6531580f5391adbe85d9c475208a56452e4c0b",
                "startedTime": "2023-09-06T08:19:17Z",
                "state": "Completed",
                "verified": false,
                "version": "4.13.9"
            }
        ],
`

Reply:
Confirmed, latest is always the [0]th item in the history...

```
[
  {
    "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.14.0-ec.4-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nForced through blocking failures: Multiple precondition checks failed:\n* Precondition \"ClusterVersionUpgrade
able\" failed because of \"AdminAckRequired\": Kubernetes 1.27 and therefore OpenShift 4.14 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958395 for details and instructions.\n* Precondition \"EtcdRecentBackup\" failed because of \"UpgradeBackupInProgress\": RecentBackup: Backup pod phase: \"P
ending\"\n* Precondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.13.11 to 4.14.0-ec.4 is unknown.",
    "completionTime": null,
    "image": "quay.io/openshift-release-dev/ocp-release:4.14.0-ec.4-x86_64",
    "startedTime": "2023-09-06T12:37:19Z",
    "state": "Partial",
    "verified": false,
    "version": "4.14.0-ec.4"
  },
  {
    "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.13.11-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nPrecondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=True (), so
 the update from 4.13.9 to 4.13.11 is probably neither recommended nor supported.",
    "completionTime": "2023-09-06T10:06:05Z",
    "image": "quay.io/openshift-release-dev/ocp-release:4.13.11-x86_64",
    "startedTime": "2023-09-06T08:52:43Z",
    "state": "Completed",
    "verified": false,
    "version": "4.13.11"
  },
  {
    "completionTime": "2023-09-06T08:52:09Z",
    "image": "quay.io/openshift-release-dev/ocp-release@sha256:a266d3d65c433b460cdef7ab5d6531580f5391adbe85d9c475208a56452e4c0b",
    "startedTime": "2023-09-06T08:19:17Z",
    "state": "Completed",
    "verified": false,
    "version": "4.13.9"
  }
]

```